src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -193,170 +202,59 @@ internal JsonTypeInfo(Type type, JsonConverter converter, JsonSerializerOptions
                     Debug.Fail($"Unexpected class type: {PropertyInfoForTypeInfo.ConverterStrategy}");
                     throw new InvalidOperationException();
             }
-
-            CreateObject = Options.MemberAccessorStrategy.CreateConstructor(type);
-
-            // These two method overrides are expected to perform
-            // orthogonal changes, so we can invoke them both safely.
-            converter.ConfigureJsonTypeInfo(this, options);
-            converter.ConfigureJsonTypeInfoUsingReflection(this, options);
         }
 
-        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        private void AddPropertiesAndParametersUsingReflection()
-        {
-            Debug.Assert(PropertyInfoForTypeInfo.ConverterStrategy == ConverterStrategy.Object);
+        private bool _isConfigured;
 
-            const BindingFlags bindingFlags =
-                BindingFlags.Instance |
-                BindingFlags.Public |
-                BindingFlags.NonPublic |
-                BindingFlags.DeclaredOnly;
-            Dictionary<string, JsonPropertyInfo>? ignoredMembers = null;
+        internal void EnsureConfigured()
+        {
+            if (_isConfigured)
```

Comment:
Given that this is method is currently subject to contention, we might need to use volatile writes and reads here.

Per my previous comments though, I believe we can guarantee that this is initialized by a single thead (e.g. by moving `EnsureConfigured` inside the `JsonSerializerOptions.GetJsonTypeInfoFromContextOrCreate` method.

Reply:
Actually I've done following things to make sure this doesn't cause any issues:
- I've manually reviewed the code paths to make sure we do signle assignment once specific stage is finished (meaning it's ok for this to be double initialized)
- I've run the tests in a loop for several hours to flush out (hopefully all) concurrency issues

Reply:
> I've manually reviewed the code paths to make sure we do signle assignment once specific stage is finished (meaning it's ok for this to be double initialized)

Might make sense to add comments to prevent any future regressions in that case. Also, this wouldn't prevent the `_isConfigured` sentinel from being reordered by the CPU or compiler.

> I've run the tests in a loop for several hours to flush out (hopefully all) concurrency issues

In my experience our CI is much better in flushing out concurrency issues versus any local stress testing

Reply:
I really doubt reordering would be as aggressive as you mention, I can see reordering happen between assigning fields but not branch and assignment

Reply:
In any case, I would try to run unit tests in a loop for a while just to be sure. (as we found out recently, it's pretty capable of surfacing issues like this -- probably better than a dedicated console app doing stress tests)

Reply:
I've already run it for around a day